# **Diabetic Readmission Predictor — Data Cleaning & Exploratory Analysis**

# **Dataset Overview**

Dataset: Diabetes 130-US Hospitals for Years 1999–2008

Source: UC Irvine Machine Learning Repository

Owner: Srimayee Bhattacharyya

Dimensions: 101,766 rows × 50 columns (before cleaning)

# **About the Dataset**

The Diabetes 130-US Hospitals Dataset contains over 100,000 hospital admissions of diabetic patients collected from 130 hospitals in the United States between 1999 and 2008.

Each record represents one hospital encounter and contains demographic information, admission details, diagnoses, laboratory procedures, medications, length of stay, and the patient's readmission status.

The primary objective of this project is to clean and prepare the dataset for exploratory analysis and predictive modeling to identify the factors associated with hospital readmission.

# **Dataset Features**

The dataset contains information related to:

1. Patient demographics
2. Age and gender
3. Race
4. Admission details
5. Hospital stay duration
6. Medical specialties
7. Laboratory procedures
8. Medications
9. Number of diagnoses
10. Diabetes medications
11. Readmission outcome

# **Column Dictionary**

| No. | Column                       | Type     | Description                                           |
| --- | ---------------------------- | -------- | ----------------------------------------------------- |
| 1   | encounter_id                 | Integer  | Unique identifier for each hospital encounter         |
| 2   | patient_nbr                  | Integer  | Unique patient identifier                             |
| 3   | race                         | Category | Patient race                                          |
| 4   | gender                       | Category | Patient gender                                        |
| 5   | age                          | Category | Patient age group                                     |
| 6   | weight                       | Category | Patient weight                                        |
| 7   | admission_type_id            | Integer  | Type of hospital admission                            |
| 8   | discharge_disposition_id     | Integer  | Patient discharge disposition                         |
| 9   | admission_source_id          | Integer  | Source of hospital admission                          |
| 10  | time_in_hospital             | Integer  | Number of hospital stay days                          |
| 11  | payer_code                   | Category | Insurance/payment code                                |
| 12  | medical_specialty            | Category | Attending physician specialty                         |
| 13  | num_lab_procedures           | Integer  | Number of laboratory procedures                       |
| 14  | num_procedures               | Integer  | Number of procedures performed                        |
| 15  | num_medications              | Integer  | Number of medications prescribed                      |
| 16  | number_outpatient            | Integer  | Previous outpatient visits                            |
| 17  | number_emergency             | Integer  | Previous emergency visits                             |
| 18  | number_inpatient             | Integer  | Previous inpatient visits                             |
| 19  | diag_1                       | Category | Primary diagnosis                                     |
| 20  | diag_2                       | Category | Secondary diagnosis                                   |
| 21  | diag_3                       | Category | Tertiary diagnosis                                    |
| 22  | number_diagnoses             | Integer  | Number of diagnoses                                   |
| 23  | max_glu_serum                | Category | Max glucose serum result                              |
| 24  | A1Cresult                    | Category | A1C test result                                       |
| 25  | metformin                    | Category | Dosage of metformin                                   |
| 26  | repaglinide                  | Category | Dosage of repaglinide                                 |
| 27  | nateglinide                  | Category | Dosage of nateglinide                                 |
| 28  | chlorpropamide               | Category | Dosage of chlorpropamide                              |
| 29  | glimepiride                  | Category | Dosage of glimepiride                                 |
| 30  | acetohexamide                | Category | Dosage of acetohexamide                               |
| 31  | glipizide                    | Category | Dosage of glipizide                                   |
| 32  | glyburide                    | Category | Dosage of glyburide                                   |
| 33  | tolbutamide                  | Category | Dosage of tolbutamide                                 |
| 34  | pioglitazone                 | Category | Dosage of pioglitazone                                |
| 35  | rosiglitazone                | Category | Dosage of rosiglitazone                               |
| 36  | acarbose                     | Category | Dosage of acarbose                                    |
| 37  | miglitol                     | Category | Dosage of miglitol                                    |
| 38  | troglitazone                 | Category | Dosage of troglitazone                                |
| 39  | tolazamide                   | Category | Dosage of tolazamide                                  |
| 40  | examide                      | Category | Dosage of examide                                     |
| 41  | citoglipton                  | Category | Dosage of citoglipton                                 |
| 42  | insulin                      | Category | Dosage of insulin                                     |
| 43  | glyburide-metformin          | Category | Dosage of glyburide-metformin                         |
| 44  | glipizide-metformin          | Category | Dosage of glipizide-metformin                         |
| 45  | glimepiride-pioglitazone     | Category | Dosage of glimepiride-pioglitazone                    |
| 46  | metformin-rosiglitazone     | Category | Dosage of metformin-rosiglitazone                      |
| 47  | metformin-pioglitazone      | Category | Dosage of metformin-pioglitazone                       |
| 48  | change                       | Category | Change in diabetic medication (Yes/No)                |
| 49  | diabetesMed                  | Category | Diabetic medication prescribed (Yes/No)               |
| 50  | readmitted                   | Category | Readmission status (<30 days, >30 days, No Readmission)|
| 51  | admission_type_description   | Category | Description of the admission type                     |
| 52  | discharge_description        | Category | Description of the discharge disposition              |
| 53  | admission_source_description | Category | Description of the admission source                   |

## Data Quality Summary (Pre-Cleaning)

| Issue                         | Details                                                                                                 |
| :---------------------------- | :------------------------------------------------------------------------------------------------------ |
| **Total rows**                | 101,766                                                                                                 |
| **Total columns**             | 53                                                                                                      |
| **Duplicate rows**            | 0                                                                                                       |
| **Columns with missing values**| 5 columns contained at least some NaN values: `max_glu_serum` (96420), `A1Cresult` (84748), `admission_type_description` (5291), `admission_source_description` (6781), `discharge_description` (3691) |
| **Largest missing-value columns**| `max_glu_serum` (96,420 NaN), `A1Cresult` (84,748 NaN)                                                     |
| **Data type distribution**      | `int64` (13), `object` (40)                                                                               |
| **Data type mismatches**      | No obvious mismatches; several categorical features (`race`, `gender`, `age`, `payer_code`, `medical_specialty`, `diag_1`, `diag_2`, `diag_3`, and medication columns) are correctly stored as `object` (string) type. Integer ID columns (`admission_type_id`, `discharge_disposition_id`, `admission_source_id`) are `int64`. |
| **Prefix noise**              | No apparent prefix noise like 'ORD-0' or 'CUST' in the original columns.                                |
| **Date format inconsistency** | No date columns were identified in the dataset.                                                         |
| **Negative values**           | No invalid negative values found in integer columns.                                                    |

## Cleaning Strategy Proposed

Based on the initial data quality assessment and domain understanding, the following data cleaning and preparation steps are proposed for this dataset:

*   **Missing Value Identification**: Identify '?' characters (e.g., in `race`, `weight`, `payer_code`, `medical_specialty`, `diag_1`, `diag_2`, `diag_3`) as missing values and replace them with `np.nan` for consistent handling across the dataset.
*   **Feature Enrichment**: Merge the `diabetic_df` with the `IDS_mapping.csv` file (if necessary, or ensure descriptions are correctly linked/available) to obtain descriptive names for `admission_type_id`, `discharge_disposition_id`, and `admission_source_id`. This enhances readability and interpretability.
*   **High Missing Value Columns**: Consider dropping columns with a very high percentage of missing values (e.g., `weight`, `max_glu_serum`, `A1Cresult`), as they may contain insufficient information to be useful for analysis or imputation. A threshold for dropping will be determined.
*   **Categorical Missing Value Imputation**: For remaining categorical columns with missing values (e.g., `admission_type_description`, `discharge_description`, `admission_source_description` and others after '?' replacement), `np.nan` values will be imputed with 'Unknown' to preserve the records and signify their unknown status. This prevents loss of potentially valuable data rows.
*   **Inconsistent Data Handling**: Address any 'Not Mapped' values (e.g., in `discharge_description`, `admission_source_description`) by treating them similarly to other missing values, replacing them with 'Unknown' or appropriate `np.nan` values.

## Analytical Goals

This project aims to achieve the following analytical objectives based on the cleaned and preprocessed dataset:

*   **Identify Key Factors** — Determine which patient demographics, admission details, medical procedures, and medication usages are most significantly associated with hospital readmission for diabetic patients.
*   **Predictive Modeling** — Develop a robust predictive model to accurately forecast the likelihood of a patient being readmitted. The model will distinguish between readmission within 30 days, readmission after 30 days, and no readmission.
*   **Insights for Intervention** — Provide actionable insights that can inform healthcare providers and policy makers on potential interventions to reduce readmission rates and improve patient outcomes.
*   **Feature Importance** — Understand the relative importance of different features in predicting readmission, highlighting areas for targeted clinical focus.

**Environment Setup**

In [1]:
import numpy as np
import matplotlib as plt
import pandas as pd
import seaborn as sns
import plotly.express as px

**Load the dataset**

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
df=pd.read_csv('/content/merged_diabetic_data.csv')

**Basic Understanding regarding the dataset**

Display the top 5 rows in the dataset

In [4]:
df.head()

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted,admission_type_description,discharge_description,admission_source_description
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,NO,NaN,Not Mapped,Physician Referral
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,No,No,No,Ch,Yes,>30,Emergency,Discharged to home,Emergency Room
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,Yes,NO,Emergency,Discharged to home,Emergency Room
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,No,No,No,Ch,Yes,NO,Emergency,Discharged to home,Emergency Room
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,No,No,No,Ch,Yes,NO,Emergency,Discharged to home,Emergency Room


Display the last 5 rows in the dataset

In [5]:
df.tail()

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted,admission_type_description,discharge_description,admission_source_description
101761,443847548,100162476,AfricanAmerican,Male,[70-80),?,1,3,7,3,...,No,No,No,No,Ch,Yes,>30,Emergency,Discharged/transferred to SNF,Emergency Room
101762,443847782,74694222,AfricanAmerican,Female,[80-90),?,1,4,5,5,...,No,No,No,No,No,Yes,NO,Emergency,Discharged/transferred to ICF,Transfer from a Skilled Nursing Facility (SNF)
101763,443854148,41088789,Caucasian,Male,[70-80),?,1,1,7,1,...,No,No,No,No,Ch,Yes,NO,Emergency,Discharged to home,Emergency Room
101764,443857166,31693671,Caucasian,Female,[80-90),?,2,3,7,10,...,No,No,No,No,Ch,Yes,NO,Urgent,Discharged/transferred to SNF,Emergency Room
101765,443867222,175429310,Caucasian,Male,[70-80),?,1,1,7,6,...,No,No,No,No,No,No,NO,Emergency,Discharged to home,Emergency Room


Read the data from 21 to 30

In [6]:
df.head(31).tail(10)

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted,admission_type_description,discharge_description,admission_source_description
21,150048,21239181,?,Male,[60-70),?,2,1,4,2,...,No,No,No,No,Ch,Yes,NO,Urgent,Discharged to home,Transfer from a hospital
22,182796,63000108,AfricanAmerican,Female,[70-80),?,2,1,4,2,...,No,No,No,No,No,No,NO,Urgent,Discharged to home,Transfer from a hospital
23,183930,107400762,Caucasian,Female,[80-90),?,2,6,1,11,...,No,No,No,No,No,No,>30,Urgent,Discharged/transferred to home with home healt...,Physician Referral
24,216156,62718876,AfricanAmerican,Female,[70-80),?,3,1,2,3,...,No,No,No,No,Ch,Yes,NO,Elective,Discharged to home,Clinic Referral
25,221634,21861756,Other,Female,[50-60),?,1,1,7,1,...,No,No,No,No,No,Yes,NO,Emergency,Discharged to home,Emergency Room
26,236316,40523301,Caucasian,Male,[80-90),?,1,3,7,6,...,No,No,No,No,Ch,Yes,NO,Emergency,Discharged/transferred to SNF,Emergency Room
27,248916,115196778,Caucasian,Female,[50-60),?,1,1,1,2,...,No,No,No,No,No,Yes,>30,Emergency,Discharged to home,Physician Referral
28,250872,41606064,Caucasian,Male,[20-30),?,2,1,2,10,...,No,No,No,No,Ch,Yes,>30,Urgent,Discharged to home,Clinic Referral
29,252822,18196434,Caucasian,Female,[80-90),?,1,2,7,5,...,No,No,No,No,Ch,Yes,>30,Emergency,Discharged/transferred to another short term h...,Emergency Room
30,253380,56480238,AfricanAmerican,Female,[60-70),?,1,1,7,6,...,No,No,No,No,Ch,Yes,NO,Emergency,Discharged to home,Emergency Room


Display first Row

In [7]:
df.iloc[[0]]

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted,admission_type_description,discharge_description,admission_source_description
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,NO,NaN,Not Mapped,Physician Referral


Display Rows from 21 to 30

In [8]:
df.iloc[21:31]

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted,admission_type_description,discharge_description,admission_source_description
21,150048,21239181,?,Male,[60-70),?,2,1,4,2,...,No,No,No,No,Ch,Yes,NO,Urgent,Discharged to home,Transfer from a hospital
22,182796,63000108,AfricanAmerican,Female,[70-80),?,2,1,4,2,...,No,No,No,No,No,No,NO,Urgent,Discharged to home,Transfer from a hospital
23,183930,107400762,Caucasian,Female,[80-90),?,2,6,1,11,...,No,No,No,No,No,No,>30,Urgent,Discharged/transferred to home with home healt...,Physician Referral
24,216156,62718876,AfricanAmerican,Female,[70-80),?,3,1,2,3,...,No,No,No,No,Ch,Yes,NO,Elective,Discharged to home,Clinic Referral
25,221634,21861756,Other,Female,[50-60),?,1,1,7,1,...,No,No,No,No,No,Yes,NO,Emergency,Discharged to home,Emergency Room
26,236316,40523301,Caucasian,Male,[80-90),?,1,3,7,6,...,No,No,No,No,Ch,Yes,NO,Emergency,Discharged/transferred to SNF,Emergency Room
27,248916,115196778,Caucasian,Female,[50-60),?,1,1,1,2,...,No,No,No,No,No,Yes,>30,Emergency,Discharged to home,Physician Referral
28,250872,41606064,Caucasian,Male,[20-30),?,2,1,2,10,...,No,No,No,No,Ch,Yes,>30,Urgent,Discharged to home,Clinic Referral
29,252822,18196434,Caucasian,Female,[80-90),?,1,2,7,5,...,No,No,No,No,Ch,Yes,>30,Emergency,Discharged/transferred to another short term h...,Emergency Room
30,253380,56480238,AfricanAmerican,Female,[60-70),?,1,1,7,6,...,No,No,No,No,Ch,Yes,NO,Emergency,Discharged to home,Emergency Room


Display 2nd to 5th rows from encounter_id

In [9]:
df.loc[2:6,"encounter_id"]

,encounter_id
2,64410
3,500364
4,16680
5,35754
6,55842


How many rows and columns are present in this dataset?

In [10]:
df.shape

(101766, 53)

Find the columns and their datatypes

In [11]:
df.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted',
       'admission_type_description', 'discharge_description',
       'admission_source_descr

In [12]:
df.dtypes

,0
encounter_id,int64
patient_nbr,int64
race,object
gender,object
age,object
weight,object
admission_type_id,int64
discharge_disposition_id,int64
admission_source_id,int64
time_in_hospital,int64


Overview of the dataset

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 53 columns):
 #   Column                        Non-Null Count   Dtype 
---  ------                        --------------   ----- 
 0   encounter_id                  101766 non-null  int64 
 1   patient_nbr                   101766 non-null  int64 
 2   race                          101766 non-null  object
 3   gender                        101766 non-null  object
 4   age                           101766 non-null  object
 5   weight                        101766 non-null  object
 6   admission_type_id             101766 non-null  int64 
 7   discharge_disposition_id      101766 non-null  int64 
 8   admission_source_id           101766 non-null  int64 
 9   time_in_hospital              101766 non-null  int64 
 10  payer_code                    101766 non-null  object
 11  medical_specialty             101766 non-null  object
 12  num_lab_procedures            101766 non-null  int64 
 13 

In [14]:
df.isnull().sum()

,0
encounter_id,0
patient_nbr,0
race,0
gender,0
age,0
weight,0
admission_type_id,0
discharge_disposition_id,0
admission_source_id,0
time_in_hospital,0


Find unique values of all columns

In [15]:
df.nunique().sort_values()

,0
citoglipton,1
examide,1
acetohexamide,2
glipizide-metformin,2
tolbutamide,2
troglitazone,2
metformin-rosiglitazone,2
glimepiride-pioglitazone,2
metformin-pioglitazone,2
change,2


**Find the duplicate rows**

In [16]:
df.duplicated().sum()

np.int64(0)

The output np.int64(0) means that there are zero duplicate rows in your DataFrame df

In [17]:
df.describe()

,encounter_id,patient_nbr,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,num_lab_procedures,num_procedures,num_medications,number_outpatient,number_emergency,number_inpatient,number_diagnoses
count,1.017660e+05,1.017660e+05,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000
mean,1.652016e+08,5.433040e+07,2.024006,3.715642,5.754437,4.395987,43.095641,1.339730,16.021844,0.369357,0.197836,0.635566,7.422607
std,1.026403e+08,3.869636e+07,1.445403,5.280166,4.064081,2.985108,19.674362,1.705807,8.127566,1.267265,0.930472,1.262863,1.933600
min,1.252200e+04,1.350000e+02,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000
25%,8.496119e+07,2.341322e+07,1.000000,1.000000,1.000000,2.000000,31.000000,0.000000,10.000000,0.000000,0.000000,0.000000,6.000000
50%,1.523890e+08,4.550514e+07,1.000000,1.000000,7.000000,4.000000,44.000000,1.000000,15.000000,0.000000,0.000000,0.000000,8.000000
75%,2.302709e+08,8.754595e+07,3.000000,4.000000,7.000000,6.000000,57.000000,2.000000,20.000000,0.000000,0.000000,1.000000,9.000000
max,4.438672e+08,1.895026e+08,8.000000,28.000000,25.000000,14.000000,132.000000,6.000000,81.000000,42.000000,76.000000,21.000000,16.000000


1. Count : Number of non-null values
2. mean : Average value
3. std : Standard deviation (spread of data)
4. min : Smallest value
5. 25% : First Quartile(Q1)
6. 50% : Median
7. 75% : Third Quartile
8. max : Largest Value

# **Cleaning Notes**

### 1. `encounter_id`
- Check for missing values.
- Verify that each encounter ID is unique.
- Convert data type to **int64** if required.

### 2. `patient_nbr`
- Check for missing values.
- Convert data type to **int64**.
- Retain duplicate patient IDs since one patient may have multiple hospital encounters.

### 3. `race`
- Replace `"?"` with **NaN**.
- Fill missing values using the **mode**.
- Standardize category names.

### 4. `gender`
- Check for `"Unknown/Invalid"` values.
- Remove or handle invalid records.
- Standardize category values.

### 5. `age`
- Verify age intervals.
- Convert to an ordered categorical variable or representative numeric values if required.

### 6. `weight`
- Replace `"?"` with **NaN**.
- Remove the column due to excessive missing values (approximately 97%).

### 7. `admission_type_id`
- Check for missing values.
- Verify integer data type.
- Validate mapping with **admission_type_description**.

### 8. `discharge_disposition_id`
- Check for missing values.
- Verify integer data type.
- Validate mapping with **discharge_description**.

### 9. `admission_source_id`
- Check for missing values.
- Verify integer data type.
- Validate mapping with **admission_source_description**.

### 10. `time_in_hospital`
- Check for missing values.
- Verify values are positive integers.
- Validate hospital stay duration.

### 11. `payer_code`
- Replace `"?"` with **NaN**.
- Remove the column due to a high percentage of missing values.

### 12. `medical_specialty`
- Replace `"?"` with **NaN**.
- If the missing percentage is too high, remove the column; otherwise replace missing values with `"Unknown"`.

### 13. `num_lab_procedures`
- Check for missing values.
- Verify non-negative integer values.

### 14. `num_procedures`
- Check for missing values.
- Verify non-negative integer values.

### 15. `num_medications`
- Check for missing values.
- Verify non-negative integer values.

### 16. `number_outpatient`
- Check for missing values.
- Verify non-negative integer values.

### 17. `number_emergency`
- Check for missing values.
- Verify non-negative integer values.

### 18. `number_inpatient`
- Check for missing values.
- Verify non-negative integer values.

### 19. `diag_1`
- Replace `"?"` with `"Unknown"` or **NaN**.
- Standardize diagnosis codes.

### 20. `diag_2`
- Replace `"?"` with `"Unknown"` or **NaN**.
- Standardize diagnosis codes.

### 21. `diag_3`
- Replace `"?"` with `"Unknown"` or **NaN**.
- Standardize diagnosis codes.

### 22. `number_diagnoses`
- Check for missing values.
- Verify integer data type.

### 23. `max_glu_serum`
- Replace missing values where required.
- Standardize categorical values (`None`, `Norm`, `>200`, `>300`).

### 24. `A1Cresult`
- Replace missing values where required.
- Standardize categorical values (`None`, `Norm`, `>7`, `>8`).

### 25. `metformin`
- Verify valid medication status values (`No`, `Steady`, `Up`, `Down`).

### 26. `repaglinide`
- Verify valid medication status values.

### 27. `nateglinide`
- Verify valid medication status values.

### 28. `chlorpropamide`
- Verify valid medication status values.

### 29. `glimepiride`
- Verify valid medication status values.

### 30. `acetohexamide`
- Verify valid medication status values.

### 31. `glipizide`
- Verify valid medication status values.

### 32. `glyburide`
- Verify valid medication status values.

### 33. `tolbutamide`
- Verify valid medication status values.

### 34. `pioglitazone`
- Verify valid medication status values.

### 35. `rosiglitazone`
- Verify valid medication status values.

### 36. `acarbose`
- Verify valid medication status values.

### 37. `miglitol`
- Verify valid medication status values.

### 38. `troglitazone`
- Verify valid medication status values.

### 39. `tolazamide`
- Verify valid medication status values.

### 40. `examide`
- Verify valid medication status values.

### 41. `citoglipton`
- Verify valid medication status values.

### 42. `insulin`
- Verify valid medication status values.

### 43. `glyburide-metformin`
- Verify valid medication status values.

### 44. `glipizide-metformin`
- Verify valid medication status values.

### 45. `glimepiride-pioglitazone`
- Verify valid medication status values.

### 46. `metformin-rosiglitazone`
- Verify valid medication status values.

### 47. `metformin-pioglitazone`
- Verify valid medication status values.

### 48. `change`
- Verify valid categories (`No`, `Ch`).

### 49. `diabetesMed`
- Verify valid categories (`Yes`, `No`).

### 50. `readmitted`
- Check for missing values.
- Verify target class labels (`NO`, `>30`, `<30`).

### 51. `admission_type_description`
- Check for missing values after merging.
- Verify consistency with `admission_type_id`.

### 52. `discharge_description`
- Check for missing values after merging.
- Verify consistency with `discharge_disposition_id`.

### 53. `admission_source_description`
- Check for missing values after merging.
- Verify consistency with `admission_source_id`.

## Final Validation
- Verify the merged dataset contains **101,766 rows and 53 columns**.
- Check for duplicate rows.
- Check for remaining missing values.
- Verify data types.
- Confirm merged description columns are correctly mapped.
- Save the cleaned dataset as **`cleaned_diabetic_data.csv`**.

# **Data Cleaning**

## 1. Cleaning `encounter_id`

The `encounter_id` column uniquely identifies each hospital encounter.

Cleaning Steps:
- Check for missing values.
- Verify uniqueness.
- Convert the data type to `int64` if necessary.

Step 1: Check Missing values

In [18]:
df["encounter_id"].isnull().sum()

np.int64(0)

Step 2:  Check uniqueness

In [19]:
df["encounter_id"].nunique()

101766

Step 3: Data type

In [20]:
df["encounter_id"].dtype

dtype('int64')

In [21]:
df["encounter_id"] = df["encounter_id"].astype("int64")

## 2. Cleaning `patient_nbr`

The `patient_nbr` column uniquely identifies each patient. A patient may have multiple hospital encounters, so duplicate patient IDs are expected.

Cleaning Steps:
- Check for missing values.
- Convert the data type to `int64`.
- Verify duplicate patient IDs.

Step 1 : Check missing values

In [22]:
df["patient_nbr"].isnull().sum()

np.int64(0)

Step 2: Check for duplicate patient IDs

In [23]:
df["patient_nbr"].duplicated().sum()

np.int64(30248)

Step 3: Check Data type and convert the data type to `int64`

In [24]:
# df["patient_nbr"].dtype
df["patient_nbr"] = df["patient_nbr"].astype("int64")

## 3. Cleaning `race`

The `race` column contains patient race information. Missing values are represented using `"?"`.

Cleaning Steps:
- Replace `"?"` with `NaN`.
- Check missing values.
- Fill missing values using the mode.

Step 1: Replace `"?"` with `NaN`

In [25]:
df['race'] = df['race'].replace('?', np.nan)

Step 2: Check Missing Values

In [26]:
df['race'].isnull().sum()

np.int64(2273)

### Step 3: Fill missing values using the mode

In [27]:
mode_race = df['race'].mode()[0]
df['race'] = df['race'].fillna(mode_race)

### Step 4: Verify that missing values have been handled and check the distribution

In [28]:
df["race"].isnull().sum()
df["race"].value_counts()

,count
race,
Caucasian,78372
AfricanAmerican,19210
Hispanic,2037
Other,1506
Asian,641


## 4. Cleaning `gender`

The `gender` column contains patient gender information. We need to handle 'Unknown/Invalid' entries.

### Step 1: Identify 'Unknown/Invalid' values

In [29]:
df['gender'].value_counts()

,count
gender,
Female,54708
Male,47055
Unknown/Invalid,3


### Step 2: Remove or handle invalid records

Given the small number of 'Unknown/Invalid' entries (only 3 in 101,766 records), it is appropriate to remove these rows to maintain data quality.

In [30]:
df = df[df["gender"] != "Unknown/Invalid"]

df.shape

df["gender"].value_counts()

,count
gender,
Female,54708
Male,47055


## 5. Cleaning `age`

The `age` column contains patient age in intervals.

Cleaning Steps:
- Verify age intervals.
- Convert to an ordered categorical variable or representative numeric values if required.

### Step 1: Verify age intervals and check their distribution

In [31]:
df['age'].value_counts().sort_index()

,count
age,
[0-10),161
[10-20),691
[20-30),1657
[30-40),3775
[40-50),9685
[50-60),17256
[60-70),22482
[70-80),26066
[80-90),17197


### Step 2: Convert age intervals to numerical representation

To make the `age` column suitable for numerical analysis, we will convert the age ranges into their midpoint values.

In [32]:
age_mapping = {
    '[0-10)': 5,
    '[10-20)': 15,
    '[20-30)': 25,
    '[30-40)': 35,
    '[40-50)': 45,
    '[50-60)': 55,
    '[60-70)': 65,
    '[70-80)': 75,
    '[80-90)': 85,
    '[90-100)': 95
}
df['age'] = df['age'].map(age_mapping)
df["age"].value_counts().sort_index()

,count
age,
5,161
15,691
25,1657
35,3775
45,9685
55,17256
65,22482
75,26066
85,17197


## 6. Cleaning `weight`

The `weight` column contains patient weight information. It is expected to have many missing values.

### Step 1: Replace '?' with `np.nan` and check missing values

In [33]:
df['weight'] = df['weight'].replace('?', np.nan)
missing_weight_count = df['weight'].isnull().sum()
total_rows = len(df)
missing_weight_percentage = (missing_weight_count / total_rows) * 100

print(f"Missing values in 'weight': {missing_weight_count} ({missing_weight_percentage:.2f}%)")

Missing values in 'weight': 98566 (96.86%)


### Step 2: Remove the column due to excessive missing values

Given that approximately 96.86% of the `weight` column's values are missing, this column provides insufficient information for meaningful analysis or imputation. Therefore, we will drop the `weight` column from the DataFrame.

In [34]:
df = df.drop(columns=["weight"])
df.shape

(101763, 52)

## 7. Cleaning `admission_type_id`

The `admission_type_id` column represents the type of admission.

### Step 1: Check for missing values

In [35]:
df["admission_type_id"].isnull().sum()

np.int64(0)

### Step 2: Verify integer data type

In [36]:
df["admission_type_id"].dtype

dtype('int64')

### Step 3: Validate mapping with `admission_type_description`

We will examine the unique values in `admission_type_id` and compare them with the corresponding `admission_type_description` to ensure consistency. This also helps identify any 'Not Mapped' or inconsistent descriptions.

In [37]:
print("Unique values and their counts for 'admission_type_id':")
print(df['admission_type_id'].value_counts().sort_index())
print("\nUnique values and their counts for 'admission_type_description':")
df[[
    "admission_type_id",
    "admission_type_description"
]].drop_duplicates().sort_values("admission_type_id")

Unique values and their counts for 'admission_type_id':
admission_type_id
1    53988
2    18480
3    18868
4       10
5     4785
6     5291
7       21
8      320
Name: count, dtype: int64

Unique values and their counts for 'admission_type_description':


,admission_type_id,admission_type_description
1,1,Emergency
5,2,Urgent
6,3,Elective
2043,4,Newborn
3089,5,Not Available
0,6,NaN
45829,7,Trauma Center
7789,8,Not Mapped


## 8. Cleaning `discharge_disposition_id`

The `discharge_disposition_id` column indicates the patient's status upon discharge.

### Step 1: Check for missing values

In [38]:
print(f"Missing values in 'discharge_disposition_id': {df['discharge_disposition_id'].isnull().sum()}")

Missing values in 'discharge_disposition_id': 0


### Step 2: Verify integer data type

In [39]:
df[[
    "discharge_disposition_id",
    "discharge_description"
]].drop_duplicates().sort_values("discharge_disposition_id")

,discharge_disposition_id,discharge_description
1,1,Discharged to home
29,2,Discharged/transferred to another short term h...
9,3,Discharged/transferred to SNF
972,4,Discharged/transferred to ICF
31,5,Discharged/transferred to another type of inpa...
13,6,Discharged/transferred to home with home healt...
82,7,Left AMA
3558,8,Discharged/transferred to home under care of H...
16783,9,Admitted as an inpatient to this hospital
487,10,Neonate discharged to another hospital for neo...


### Step 3: Validate mapping with `discharge_description`

We will examine the unique values in `discharge_disposition_id` and compare them with the corresponding `discharge_description` to ensure consistency.

In [40]:
print("Unique values and their counts for 'discharge_disposition_id':")
print(df['discharge_disposition_id'].value_counts().sort_index())
print("\nUnique values and their counts for 'discharge_description':")
df['discharge_description'].value_counts()

Unique values and their counts for 'discharge_disposition_id':
discharge_disposition_id
1     60232
2      2128
3     13954
4       815
5      1184
6     12902
7       623
8       108
9        21
10        6
11     1642
12        3
13      399
14      372
15       63
16       11
17       14
18     3691
19        8
20        2
22     1992
23      412
24       48
25      989
27        5
28      139
Name: count, dtype: int64

Unique values and their counts for 'discharge_description':


,count
discharge_description,
Discharged to home,60232
Discharged/transferred to SNF,13954
Discharged/transferred to home with home health service,12902
Discharged/transferred to another short term hospital,2128
Discharged/transferred to another rehab fac including rehab units of a hospital .,1992
Expired,1642
Discharged/transferred to another type of inpatient care institution,1184
Not Mapped,989
Discharged/transferred to ICF,815


## 9. Cleaning `admission_source_id`

The `admission_source_id` column indicates the source from which the patient was admitted to the hospital.

### Step 1: Check for missing values

In [41]:
df["admission_source_id"].isnull().sum()

np.int64(0)

### Step 2: Verify integer data type

In [42]:
df["admission_source_id"].dtype

dtype('int64')

### Step 3: Validate mapping with `admission_source_description`

We will examine the unique values in `admission_source_id` and compare them with the corresponding `admission_source_description` to ensure consistency.

In [43]:
df[[
    "admission_source_id",
    "admission_source_description"
]].drop_duplicates().sort_values("admission_source_id")
df['admission_source_description'].value_counts()

,count
admission_source_description,
Emergency Room,57492
Physician Referral,29564
Transfer from a hospital,3187
Transfer from another health care facility,2264
Clinic Referral,1104
Transfer from a Skilled Nursing Facility (SNF),855
HMO Referral,187
Not Mapped,161
Not Available,125


### Step 4: Handle irrelevant discharge dispositions

Certain `discharge_disposition_id` values indicate outcomes that are not considered true readmission events for our prediction task (e.g., patient expired, transferred to another facility, or internal transfers within the same hospital). We will remove these records from the dataset.

The IDs to remove are:
- **9**: 'Admitted as an inpatient to this hospital'
- **11, 19, 20**: 'Expired' (various types)
- **18**: Likely corresponds to missing/NaN descriptions, representing unclear discharge
- **25**: 'Not Mapped'

### Step 4: Identify and handle 'Not Mapped' or 'Not Available' admission sources

Based on the unique values, we observe descriptions like ' Not Mapped' and ' Not Available'. These represent admission sources that are either undefined or not relevant for direct analysis. We will identify the `admission_source_id` values associated with these descriptions and either remove them or group them into a single 'Unknown' category to streamline the data.

In [44]:
print("Unique admission_source_ids and their corresponding descriptions:")
# Create a mapping from ID to description for easier lookup
id_to_description = df.groupby('admission_source_id')['admission_source_description'].apply(lambda x: x.mode()[0] if not x.mode().empty else None)

for id_val, desc_val in id_to_description.items():
    print(f"ID: {id_val}, Description: {desc_val}")

# Identify IDs that correspond to 'Not Mapped' or 'Not Available' descriptions
# Note: Stripping whitespace from descriptions due to observed leading spaces
ids_to_reclassify = df[df['admission_source_description'].isin([' Not Mapped', ' Not Available'])]['admission_source_id'].unique()

print(f"\nAdmission Source IDs corresponding to 'Not Mapped' or 'Not Available': {ids_to_reclassify}")

Unique admission_source_ids and their corresponding descriptions:
ID: 1, Description:  Physician Referral
ID: 2, Description: Clinic Referral
ID: 3, Description: HMO Referral
ID: 4, Description: Transfer from a hospital
ID: 5, Description:  Transfer from a Skilled Nursing Facility (SNF)
ID: 6, Description:  Transfer from another health care facility
ID: 7, Description:  Emergency Room
ID: 8, Description:  Court/Law Enforcement
ID: 9, Description:  Not Available
ID: 10, Description:  Transfer from critial access hospital
ID: 11, Description: Normal Delivery
ID: 13, Description:  Sick Baby
ID: 14, Description:  Extramural Birth
ID: 17, Description: None
ID: 20, Description:  Not Mapped
ID: 22, Description:  Transfer from hospital inpt/same fac reslt in a sep claim
ID: 25, Description:  Transfer from Ambulatory Surgery Center

Admission Source IDs corresponding to 'Not Mapped' or 'Not Available': [20  9]


### Step 5: Consolidate 'Not Mapped', 'Not Available' and 'None' descriptions into an 'Unknown' category

We will replace `admission_source_id` values 9, 17, and 20 with a new ID (e.g., 26) that will represent an 'Unknown' admission source. This ensures consistency and better interpretability for our analysis. After this, we will update the `admission_source_description` column to reflect these changes.

In [45]:
# Replace admission_source_id 9, 17, and 20 with a new ID for 'Unknown'
ids_to_replace = [9, 17, 20]
unknown_id = 26 # Choose an ID not currently in use
df['admission_source_id'] = df['admission_source_id'].replace(ids_to_replace, unknown_id)

# Update the admission_source_description column to reflect 'Unknown' for the new ID
df.loc[df['admission_source_id'] == unknown_id, 'admission_source_description'] = 'Unknown'

print("Value counts for 'admission_source_id' after consolidation:")
df['admission_source_id'].value_counts().sort_index()

print("\nValue counts for 'admission_source_description' after consolidation:")
df['admission_source_description'].value_counts()

Value counts for 'admission_source_id' after consolidation:

Value counts for 'admission_source_description' after consolidation:


,count
admission_source_description,
Emergency Room,57492
Physician Referral,29564
Unknown,7067
Transfer from a hospital,3187
Transfer from another health care facility,2264
Clinic Referral,1104
Transfer from a Skilled Nursing Facility (SNF),855
HMO Referral,187
Court/Law Enforcement,16


## 10. Cleaning `time_in_hospital`

The `time_in_hospital` column records the number of days the patient stayed in the hospital.

### Step 1: Check for missing values

In [46]:
print(f"Missing values in 'time_in_hospital': {df['time_in_hospital'].isnull().sum()}")

Missing values in 'time_in_hospital': 0


### Step 2: Verify integer data type

In [47]:
print(f"Data type of 'time_in_hospital': {df['time_in_hospital'].dtype}")

Data type of 'time_in_hospital': int64


### Step 3: Validate values

In [48]:
print("Unique values and their counts for 'time_in_hospital':")
print(df['time_in_hospital'].value_counts().sort_index())

Unique values and their counts for 'time_in_hospital':
time_in_hospital
1     14206
2     17224
3     17756
4     13924
5      9966
6      7539
7      5859
8      4390
9      3002
10     2342
11     1855
12     1448
13     1210
14     1042
Name: count, dtype: int64


## 11. Cleaning `payer_code`

The `payer_code` column contains information about the patient's insurance or payment plan.

### Step 1: Replace '?' with `np.nan` and check missing values

In [49]:
df['payer_code'] = df['payer_code'].replace('?', np.nan)
missing_payer_code_count = df['payer_code'].isnull().sum()
total_rows = len(df)
missing_payer_code_percentage = (missing_payer_code_count / total_rows) * 100

print(f"Missing values in 'payer_code': {missing_payer_code_count} ({missing_payer_code_percentage:.2f}%)")

Missing values in 'payer_code': 40255 (39.56%)


### Step 2: Remove the column due to excessive missing values

Given the high percentage of missing values, the `payer_code` column is unlikely to contribute meaningfully to our analysis. Therefore, we will drop it.

In [50]:
df = df.drop('payer_code', axis=1)
print(f"New shape of DataFrame after dropping 'payer_code' column: {df.shape}")

New shape of DataFrame after dropping 'payer_code' column: (101763, 51)


## 12. Cleaning `medical_specialty`

The `medical_specialty` column indicates the specialty of the attending physician. This column may have missing values represented by '?'.

### Step 1: Replace '?' with `np.nan` and check missing values

In [51]:
df['medical_specialty'] = df['medical_specialty'].replace('?', np.nan)
missing_medical_specialty_count = df['medical_specialty'].isnull().sum()
total_rows = len(df)
missing_medical_specialty_percentage = (missing_medical_specialty_count / total_rows) * 100

print(f"Missing values in 'medical_specialty': {missing_medical_specialty_count} ({missing_medical_specialty_percentage:.2f}%)")

Missing values in 'medical_specialty': 49947 (49.08%)


### Step 2: Handle missing values based on percentage

If the percentage of missing values is very high, we might consider dropping the column. Otherwise, we will fill the missing values with 'Unknown'.

In [52]:
if missing_medical_specialty_percentage > 50: # Define a threshold, e.g., 50%
    df = df.drop('medical_specialty', axis=1)
    print(f"'medical_specialty' column dropped due to {missing_medical_specialty_percentage:.2f}% missing values. New shape: {df.shape}")
else:
    df['medical_specialty'] = df['medical_specialty'].fillna('Unknown')
    print("Missing values in 'medical_specialty' filled with 'Unknown'.")
    print("Value counts for 'medical_specialty' after handling:")
    print(df['medical_specialty'].value_counts())

Missing values in 'medical_specialty' filled with 'Unknown'.
Value counts for 'medical_specialty' after handling:
medical_specialty
Unknown                   49947
InternalMedicine          14635
Emergency/Trauma           7565
Family/GeneralPractice     7440
Cardiology                 5351
                          ...  
Proctology                    1
Speech                        1
SportsMedicine                1
Perinatology                  1
Neurophysiology               1
Name: count, Length: 73, dtype: int64


## 13. Cleaning `num_lab_procedures`

The `num_lab_procedures` column records the number of laboratory procedures performed during the encounter.

### Step 1: Check for missing values

In [53]:
print(f"Missing values in 'num_lab_procedures': {df['num_lab_procedures'].isnull().sum()}")

Missing values in 'num_lab_procedures': 0


### Step 2: Verify non-negative integer values and data type

In [54]:
print(f"Data type of 'num_lab_procedures': {df['num_lab_procedures'].dtype}")
print("Minimum value in 'num_lab_procedures':", df['num_lab_procedures'].min())
print("Maximum value in 'num_lab_procedures':", df['num_lab_procedures'].max())

Data type of 'num_lab_procedures': int64
Minimum value in 'num_lab_procedures': 1
Maximum value in 'num_lab_procedures': 132


## 14. Cleaning `num_procedures`

The `num_procedures` column records the number of non-laboratory procedures performed during the encounter.

### Step 1: Check for missing values

In [55]:
print(f"Missing values in 'num_procedures': {df['num_procedures'].isnull().sum()}")

Missing values in 'num_procedures': 0


### Step 2: Verify non-negative integer values and data type

In [56]:
print(f"Data type of 'num_procedures': {df['num_procedures'].dtype}")
print("Minimum value in 'num_procedures':", df['num_procedures'].min())
print("Maximum value in 'num_procedures':", df['num_procedures'].max())

Data type of 'num_procedures': int64
Minimum value in 'num_procedures': 0
Maximum value in 'num_procedures': 6


## 15. Cleaning `num_medications`

The `num_medications` column records the number of different medications administered during the encounter.

### Step 1: Check for missing values

In [57]:
print(f"Missing values in 'num_medications': {df['num_medications'].isnull().sum()}")

Missing values in 'num_medications': 0


### Step 2: Verify non-negative integer values and data type

In [58]:
print(f"Data type of 'num_medications': {df['num_medications'].dtype}")
print("Minimum value in 'num_medications':", df['num_medications'].min())
print("Maximum value in 'num_medications':", df['num_medications'].max())

Data type of 'num_medications': int64
Minimum value in 'num_medications': 1
Maximum value in 'num_medications': 81


## 16. Cleaning `number_outpatient`

The `number_outpatient` column records the number of outpatient visits a patient had in the year prior to the encounter.

### Step 1: Check for missing values

In [59]:
print(f"Missing values in 'number_outpatient': {df['number_outpatient'].isnull().sum()}")

Missing values in 'number_outpatient': 0


### Step 2: Verify non-negative integer values and data type

In [60]:
print(f"Data type of 'number_outpatient': {df['number_outpatient'].dtype}")
print("Minimum value in 'number_outpatient':", df['number_outpatient'].min())
print("Maximum value in 'number_outpatient':", df['number_outpatient'].max())

Data type of 'number_outpatient': int64
Minimum value in 'number_outpatient': 0
Maximum value in 'number_outpatient': 42


## 17. Cleaning `number_emergency`

The `number_emergency` column records the number of emergency visits a patient had in the year prior to the encounter.

### Step 1: Check for missing values

In [61]:
print(f"Missing values in 'number_emergency': {df['number_emergency'].isnull().sum()}")

Missing values in 'number_emergency': 0


### Step 2: Verify non-negative integer values and data type

In [62]:
print(f"Data type of 'number_emergency': {df['number_emergency'].dtype}")
print("Minimum value in 'number_emergency':", df['number_emergency'].min())
print("Maximum value in 'number_emergency':", df['number_emergency'].max())

Data type of 'number_emergency': int64
Minimum value in 'number_emergency': 0
Maximum value in 'number_emergency': 76


## 18. Cleaning `number_inpatient`

The `number_inpatient` column records the number of inpatient visits a patient had in the year prior to the encounter.

### Step 1: Check for missing values

In [63]:
print(f"Missing values in 'number_inpatient': {df['number_inpatient'].isnull().sum()}")

Missing values in 'number_inpatient': 0


### Step 2: Verify non-negative integer values and data type

In [64]:
print(f"Data type of 'number_inpatient': {df['number_inpatient'].dtype}")
print("Minimum value in 'number_inpatient':", df['number_inpatient'].min())
print("Maximum value in 'number_inpatient':", df['number_inpatient'].max())

Data type of 'number_inpatient': int64
Minimum value in 'number_inpatient': 0
Maximum value in 'number_inpatient': 21


## 19. Cleaning `diag_1`

The `diag_1` column records the primary diagnosis of the patient. This column may contain '?' values indicating missing information.

### Step 1: Replace '?' with `np.nan` and check for missing values

In [65]:
df['diag_1'] = df['diag_1'].replace('?', np.nan)
missing_diag1_count = df['diag_1'].isnull().sum()
total_rows = len(df)
missing_diag1_percentage = (missing_diag1_count / total_rows) * 100

print(f"Missing values in 'diag_1': {missing_diag1_count} ({missing_diag1_percentage:.2f}%)")

Missing values in 'diag_1': 21 (0.02%)


### Step 2: Verify data type and check unique values

In [66]:
print(f"Data type of 'diag_1': {df['diag_1'].dtype}")
print("Number of unique values in 'diag_1':", df['diag_1'].nunique())

Data type of 'diag_1': object
Number of unique values in 'diag_1': 716


### Step 3: Impute missing values with 'Unknown'

Given the small percentage of missing values (0.02%), we will impute them with 'Unknown' to retain these records while clearly marking their unknown primary diagnosis.

In [67]:
df['diag_1'] = df['diag_1'].fillna('Unknown')

print(f"Missing values in 'diag_1' after imputation: {df['diag_1'].isnull().sum()}")
print("Value counts for 'diag_1' after imputation:")
df['diag_1'].value_counts().head() # Display top 5 for brevity

Missing values in 'diag_1' after imputation: 0
Value counts for 'diag_1' after imputation:


,count
diag_1,
428,6862
414,6580
786,4016
410,3614
486,3508


## 20. Cleaning `diag_2`

The `diag_2` column records the secondary diagnosis of the patient. This column may also contain '?' values indicating missing information.

### Step 1: Replace '?' with `np.nan` and check for missing values

In [68]:
df['diag_2'] = df['diag_2'].replace('?', np.nan)
missing_diag2_count = df['diag_2'].isnull().sum()
total_rows = len(df)
missing_diag2_percentage = (missing_diag2_count / total_rows) * 100

print(f"Missing values in 'diag_2': {missing_diag2_count} ({missing_diag2_percentage:.2f}%)")

Missing values in 'diag_2': 358 (0.35%)


### Step 2: Verify data type and check unique values

In [69]:
print(f"Data type of 'diag_2': {df['diag_2'].dtype}")
print("Number of unique values in 'diag_2':", df['diag_2'].nunique())

Data type of 'diag_2': object
Number of unique values in 'diag_2': 748


### Step 3: Impute missing values with 'Unknown'

Given the small percentage of missing values (0.36%), we will impute them with 'Unknown' to retain these records while clearly marking their unknown secondary diagnosis.

In [70]:
df['diag_2'] = df['diag_2'].fillna('Unknown')

print(f"Missing values in 'diag_2' after imputation: {df['diag_2'].isnull().sum()}")
print("Value counts for 'diag_2' after imputation:")
df['diag_2'].value_counts().head() # Display top 5 for brevity

Missing values in 'diag_2' after imputation: 0
Value counts for 'diag_2' after imputation:


,count
diag_2,
276,6752
428,6662
250,6071
427,5036
401,3736


## 21. Cleaning `diag_3`

The `diag_3` column records the tertiary diagnosis of the patient. This column may also contain '?' values indicating missing information.

### Step 1: Replace '?' with `np.nan` and check for missing values

In [71]:
df['diag_3'] = df['diag_3'].replace('?', np.nan)
missing_diag3_count = df['diag_3'].isnull().sum()
total_rows = len(df)
missing_diag3_percentage = (missing_diag3_count / total_rows) * 100

print(f"Missing values in 'diag_3': {missing_diag3_count} ({missing_diag3_percentage:.2f}%)")

Missing values in 'diag_3': 1423 (1.40%)


### Step 2: Verify data type and check unique values

In [72]:
print(f"Data type of 'diag_3': {df['diag_3'].dtype}")
print("Number of unique values in 'diag_3':", df['diag_3'].nunique())

Data type of 'diag_3': object
Number of unique values in 'diag_3': 789


In [73]:
df['diag_3'] = df['diag_3'].fillna('Unknown')

print(f"Missing values in 'diag_3' after imputation: {df['diag_3'].isnull().sum()}")
print("Value counts for 'diag_3' after imputation:")
df['diag_3'].value_counts().head() # Display top 5 for brevity

Missing values in 'diag_3' after imputation: 0
Value counts for 'diag_3' after imputation:


,count
diag_3,
250,11555
401,8288
276,5175
428,4577
427,3955


## 22. Cleaning `number_diagnoses`

The `number_diagnoses` column records the total number of diagnoses entered for the patient. We need to verify its data type and ensure there are no missing values.

### Step 1: Check for missing values

In [74]:
print(f"Missing values in 'number_diagnoses': {df['number_diagnoses'].isnull().sum()}")

Missing values in 'number_diagnoses': 0


### Step 2: Verify integer data type and value range

In [75]:
print(f"Data type of 'number_diagnoses': {df['number_diagnoses'].dtype}")
print("Minimum value in 'number_diagnoses':", df['number_diagnoses'].min())
print("Maximum value in 'number_diagnoses':", df['number_diagnoses'].max())

Data type of 'number_diagnoses': int64
Minimum value in 'number_diagnoses': 1
Maximum value in 'number_diagnoses': 16


### Step 3: Impute missing values with 'Unknown'

Given the small percentage of missing values (1.41%), we will impute them with 'Unknown' to retain these records while clearly marking their unknown tertiary diagnosis.

In [76]:
df['diag_2'] = df['diag_2'].fillna('Unknown')

print(f"Missing values in 'diag_2' after imputation: {df['diag_2'].isnull().sum()}")
print("Value counts for 'diag_2' after imputation:")
df['diag_2'].value_counts().head() # Display top 5 for brevity

Missing values in 'diag_2' after imputation: 0
Value counts for 'diag_2' after imputation:


,count
diag_2,
276,6752
428,6662
250,6071
427,5036
401,3736


## Final Step: Save Cleaned Dataset

After all the cleaning steps, the processed DataFrame will be saved to a new CSV file named `cleaned_diabetic_data.csv` for future use in exploratory analysis and predictive modeling.

In [77]:
df.to_csv('cleaned_diabetic_data.csv', index=False)
print("Cleaned data saved to 'cleaned_diabetic_data.csv'")

Cleaned data saved to 'cleaned_diabetic_data.csv'
